src/message.cr

```diff
@@ -18,15 +17,88 @@ module Crylog
     getter datetime : Time
 
     # Any extra metadata added by a `Crylog::Processors::LogProcessor`.
-    property extra : Hash(String, Crylog::Context)
+    setter extra : Hash(String, Crylog::Context)?
 
-    # Any extra metadata added when the message was logged.
-    getter context : Hash(String, Crylog::Context)
+    @context : Hash(String, Crylog::Context)?
 
     # Represents the formatted version of this message.
     @[JSON::Field(ignore: true)]
     property formatted : String = ""
 
-    def initialize(@message : String, @context : Hash(String, Crylog::Context), @severity : Crylog::Severity, @channel : String, @datetime : Time, @extra : Hash(String, Crylog::Context)); end
+    # Lambda held for lazy message evaluation.
+    @[JSON::Field(ignore: true)]
+    @block : -> Logger::MsgType
+
+    @[JSON::Field(ignore: true)]
+    @set_message_context = false
+
+    def initialize(@severity : Crylog::Severity, @channel : String, &block : -> Logger::MsgType)
+      @datetime = Time.utc
+      @block = block
+    end
+
+    # The message that was logged.
+    def message : String
+      set_message_context
+      @message || ""
+    end
+
+    # Any extra metadata added when the message was logged.
+    def context : Hash(String, Crylog::Context)
+      set_message_context
+      @context.not_nil!
+    end
+
+    def extra
+      @extra ||= Hash(String, Crylog::Context).new
+    end
+
+    def message?
```

Comment:
That may double the number of conditionals per variable if there's 1 formatter and there's still the empty context condition I mentioned above to consider.

```
# First run
getter(context) {
  # Always false
  generated if @context return @context

  # Calls block once which sets @message and @context
  # @context may be assigned nil
  # 2nd conditional check - REDUNDANT
  # The condition must be checked in set_message_context because `\.message` may
  # have been called first and the block should resolve once.
  set_message_context
  @context.not_nil!
}
```

I think a getter block would add an implicit always false conditional.  If that's not a concern I'll change it, but I'm also not sure 
```
getter(foo) {
}
```

ends up being shorter or clearer than 
```
def foo
end
```

with the exact same code in the middle.

Reply:
Hmm fair enough.  Maybe we go this route then:

```cr
def message? : Bool
  set_message_context
  return false unless msg = @message
  !msg.empty?
end
```

I think that's a bit cleaner than a whole if/else block for something so simple.  Also would like these methods to have return types defined.  Otherwise, yea this'll be fine.